src/routes/videoPlaybackProxy.ts

```diff
@@ -92,47 +94,88 @@ videoPlaybackProxy.get("/", async (c) => {
     // of this RFC, that's why `i < 5`
     for (let i = 0; i < 5; i++) {
-        const googlevideoResponse: Response = await fetchClient.call(
-            undefined,
-            location,
-            {
-                method: "HEAD",
-                headers: headersToSend,
-                redirect: "manual",
-            },
-        );
+        const googlevideoResponse: Response = await fetchClient(location, {
+            method: "HEAD",
+            headers: headersToSend,
+            redirect: "manual",
+        });
         if (googlevideoResponse.headers.has("Location")) {
             location = googlevideoResponse.headers.get("Location") as string;
             continue;
         } else {
+            headResponse = googlevideoResponse;
             break;
         }
     }
+    if (headResponse === undefined) {
+        throw new HTTPException(502, {
+            res: new Response(
+                "Google headResponse redirected too many times",
+            ),
+        });
+    }
 
-    const googlevideoResponse = await fetchClient.call(
-        undefined,
-        location,
-        {
+    // =================== REQUEST CHUNKING =======================
+    // if the requested response is larger than the chunkSize, break up the response
+    // into chunks and stream the response back to the client to avoid rate limiting
+    const { readable, writable } = new TransformStream();
+    const stream = new StreamingApi(writable, readable);
+    const googleVideoUrl = new URL(location);
+    const getChunk = async (start: number, end: number) => {
+        googleVideoUrl.searchParams.set(
+            "range",
+            `${start}-${end}`,
+        );
+        const postResponse = await fetchClient(googleVideoUrl, {
             method: "POST",
             body: new Uint8Array([0x78, 0]), // protobuf: { 15: 0 } (no idea what it means but this is what YouTube uses),
             headers: headersToSend,
-        },
+        });
+        await stream.pipe(postResponse.body);
+    };
+
+    const chunkSize = 5_000_000;
+    const totalBytes = Number(
+        headResponse.headers.get("Content-Length") || "0",
     );
 
+    // if no range sent, the client wants thw whole file, i.e. for downloads
+    const wholeRequestStartByte = Number(firstByte || "0");
+    const wholeRequestEndByte = wholeRequestStartByte + Number(totalBytes) - 1;
+
+    let chunk = Promise.resolve();
+    for (
+        let startByte = wholeRequestStartByte;
+        startByte < wholeRequestEndByte;
+        startByte += chunkSize
+    ) {
+        // i.e.
+        // 0 - 4_999_999, then
+        // 5_000_000 - 9_999_999, then
+        // 10_000_000 - 14_999_999
+        let endByte = startByte + chunkSize - 1;
+        if (endByte > wholeRequestEndByte) {
+            endByte = wholeRequestEndByte;
+        }
+        chunk = chunk.then(() => getChunk(startByte, endByte));
+    }
+    chunk.catch((err) => console.log({ err }));
```

Comment:
I need to figure out some valid error handling here, but it shouldn't crash anything if a stream breaks.

Reply:
https://github.com/iv-org/invidious-companion/pull/95/commits/be52b67b8d81be4a51b8cd19a0a977ff0a4a8a31

Aborting now. I'm not sure about logging out - when a user just closes the connection the `catch` is just `err = undefined`, so not much info to gain there. Maybe if there's actually an error I could log something?